lib/ex_doc/formatter/html/autolink.ex

```diff
@@ -370,112 +378,308 @@ defmodule ExDoc.Formatter.HTML.Autolink do
+  # - `:extension` - Default value is `".html"`
+  # - `:lib_dirs`
+  # - `:locals` - A list of local functions
+  # - `:module_id` - Module of the current doc. Default value is `nil`
+  # - `:modules_refs` - List of modules available
+  #
+  # Internal options:
+  # - `:preprocess?` - `true` or `false`. Do preprocessing and postprocessing, such as replacing backticks
+  #                     with a token
+  @doc false
+  @spec link(String.t(), language, kind, keyword) :: String.t()
+  def link(string, language, kind, options) do
+    options = Keyword.put_new(options, :preprocess?, true)
+
+    string
+    |> link_process(:preprocess, options[:preprocess?])
+    |> link(language, kind, :custom, options)
+    |> link(language, kind, :normal, options)
+    |> link_process(:postprocess, options[:preprocess?])
+  end
+
+  @doc false
+  @spec link(String.t(), language, link_type, kind, keyword) :: String.t()
+  def link(string, language, kind, link_type, options) do
+    regex = regex_link_type(language, kind, link_type)
+    replace_fun = replace_fun(language, kind, link_type, options)
+
+    result = link_process(string, :preprocess, options[:preprocess?])
+    result = Regex.replace(regex, result, replace_fun)
+    result = link_process(result, :postprocess, options[:preprocess?])
+
+    # output([before: string, after: result], options[:module_id], [Version, "Version"])
+    result
+  end
+
+  @doc false
+  def link_everything(string, options) when is_list(options) do
+    # disable preprocess every time we run link/4,
+    # and transform the string manually before and after Enum.reduce
+    options = Keyword.put_new(options, :preprocess?, false)
+
+    string = preprocess(string)
+
+    string =
+      Enum.reduce(@regexes, string, fn %{
+                                         kind: kind,
+                                         language: language,
+                                         link_type: link_type
+                                       },
+                                       acc ->
+        link(acc, language, kind, link_type, options)
+      end)
+
+    postprocess(string)
   end
 
-  defp replace_normal_links(bin, project_funs, extension, lib_dirs) do
-    Regex.replace(@normal_re, bin, fn all, match ->
-      replacement(all, match, project_funs, extension, lib_dirs)
+  defp link_process(string, _, false),
+    do: string
+
+  defp link_process(string, :preprocess, true),
+    do: preprocess(string)
+
+  defp link_process(string, :postprocess, true),
+    do: postprocess(string)
+
+  defp output(term, module_id, module_id) do
+    IO.puts(inspect(term))
+    exit(:output)
+    term
+  end
+
+  # used temporarily for debugging
+  defp output(term, module_id, print_only_module_list) do
+    if module_id in print_only_module_list do
+      IO.puts(inspect(term))
+      IO.puts("****************")
+    end
+
+    term
+  end
+
+  @doc false
+  # Replaces all backticks inside the text of custom links with @backtick_token.
```

Comment:
Did we do pre-processing before? Why do we need it now?

Reply:
We didn't. But it fixes the bug of linking inside the custom text. Before it was only working if either the opening backtick is touching the opening bracket, or the closing backtick is touching the closing bracket. 

This wouldn't pass before IIRC.
```elixir
      assert Autolink.erlang_functions("[the `:erlang.apply/2` function](other.html)") ==
               "[the `:erlang.apply/2` function](other.html)"
```

So this behaviour was inconsistent. I think it worked for local and Elixir functions and it didn't work for the rest